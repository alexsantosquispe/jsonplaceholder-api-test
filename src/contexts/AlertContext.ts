import { createContext } from 'react';

import { ALERT_TYPES } from '../constants';

export type AlertType = {
  id: string;
  title: string;
  message: string;
  type?: (typeof ALERT_TYPES)[keyof typeof ALERT_TYPES];
  timeout?: number;
};

interface AlertContextType {
  alerts: AlertType[];
  addAlert: (alert: AlertType) => void;
  removeAlert: (alertId: string) => void;
}

const AlertContext = createContext<AlertContextType | null>(null);

export default AlertContext;
