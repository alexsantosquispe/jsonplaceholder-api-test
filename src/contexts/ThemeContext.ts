import { createContext } from 'react';

export const Theme = {
  light: 'light',
  dark: 'dark'
} as const;

export type ThemeType = keyof typeof Theme;

interface ThemeContextType {
  theme: ThemeType;
  toggleTheme: () => void;
}

const ThemeContext = createContext<ThemeContextType | null>(null);

export default ThemeContext;
