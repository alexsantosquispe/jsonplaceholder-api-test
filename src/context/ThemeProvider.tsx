import { ReactNode, useState } from 'react';

import ThemeContext, { ThemeType } from './ThemeContext';

interface ThemeProviderProps {
  children: ReactNode;
}

export const ThemeProvider = ({ children }: ThemeProviderProps) => {
  const [theme, setTheme] = useState<ThemeType>('light');
  const toggleTheme = () => {
    setTheme((prev) => (prev === 'light' ? 'dark' : 'light'));
  };

  return (
    <ThemeContext.Provider value={{ theme, toggleTheme }}>
      {children}
    </ThemeContext.Provider>
  );
};
