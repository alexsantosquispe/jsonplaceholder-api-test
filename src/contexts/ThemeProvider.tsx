import { ReactNode, useEffect, useState } from 'react';

import ThemeContext, { Theme, ThemeType } from './ThemeContext';
import { THEME_KEY } from '../constants';
import {
  applyThemeToDocument,
  getStoredTheme,
  getSystemTheme,
  setLSValue
} from '../utils';

interface Props {
  children: ReactNode;
}

export const ThemeProvider = ({ children }: Props) => {
  const initialTheme = getStoredTheme() ?? getSystemTheme();
  const [currentTheme, setCurrentTheme] = useState<ThemeType>(initialTheme);

  useEffect(() => {
    applyThemeToDocument(currentTheme);
    setLSValue(THEME_KEY, currentTheme);
  }, [currentTheme]);

  useEffect(() => {
    const storedTheme = getStoredTheme();
    if (storedTheme) return;

    const mediaQuery = window.matchMedia('(prefers-color-scheme: dark)');

    const handleChange = (e: MediaQueryListEvent) => {
      setCurrentTheme(e.matches ? Theme.dark : Theme.light);
    };

    mediaQuery.addEventListener('change', handleChange);
    return () => mediaQuery.removeEventListener('change', handleChange);
  }, []);

  const toggleTheme = () => {
    setCurrentTheme((prev) => (prev === Theme.dark ? Theme.light : Theme.dark));
  };

  return (
    <ThemeContext.Provider value={{ theme: currentTheme, toggleTheme }}>
      {children}
    </ThemeContext.Provider>
  );
};
