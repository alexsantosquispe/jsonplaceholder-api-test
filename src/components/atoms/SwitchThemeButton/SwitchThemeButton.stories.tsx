import type { Meta, StoryObj } from '@storybook/react';

import { SwitchThemeButton } from './SwitchThemeButton';
import { ThemeProvider } from '../../../contexts/ThemeProvider';
import { EndpointTitle } from '../EndpointTitle/EndpointTitle';

const meta: Meta<typeof SwitchThemeButton> = {
  title: 'Atoms/SwitchThemeButton',
  component: SwitchThemeButton,
  decorators: [
    (Story) => (
      <ThemeProvider>
        <div className="dark:bg-primary flex h-[30rem] w-full items-center justify-center bg-white">
          <div className="flex flex-col items-center gap-4">
            <EndpointTitle label="Switch theme button" />
            <Story />
          </div>
        </div>
      </ThemeProvider>
    )
  ]
};

export default meta;

type Story = StoryObj<typeof SwitchThemeButton>;

export const Default: Story = {
  args: {
    classNameContainer: '',
    classNameIcon: ''
  }
};
