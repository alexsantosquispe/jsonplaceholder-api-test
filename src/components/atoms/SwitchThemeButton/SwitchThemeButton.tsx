import { memo } from 'react';
import { twMerge } from 'tailwind-merge';

import { Theme } from '../../../context/ThemeContext';
import { useTheme } from '../../../hooks/useTheme';
import { MoonIcon, SunIcon } from '../../../icons';

interface SwitchThemeButtonProps {
  classNameContainer?: string;
  classNameIcon?: string;
}

const SwitchThemeButtonComponent = ({
  classNameContainer = '',
  classNameIcon = ''
}: SwitchThemeButtonProps) => {
  const { theme, toggleTheme } = useTheme();

  return (
    <button
      aria-label={`${theme === Theme.light ? 'dark mode icon' : 'light mode icon'}`}
      onClick={toggleTheme}
      className={twMerge(
        'transition-color cursor-pointer rounded-md p-4 text-black duration-150 ease-in-out hover:bg-gray-100 md:p-2 dark:text-white dark:hover:bg-white/10',
        classNameContainer
      )}
    >
      {theme === Theme.light ? (
        <MoonIcon className={twMerge('size-5', classNameIcon)} />
      ) : (
        <SunIcon className={twMerge('size-5', classNameIcon)} />
      )}
    </button>
  );
};

export const SwitchThemeButton = memo(SwitchThemeButtonComponent);
