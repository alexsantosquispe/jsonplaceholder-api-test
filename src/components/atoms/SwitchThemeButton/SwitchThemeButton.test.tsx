import { fireEvent, render, screen } from '@testing-library/react';

import { SwitchThemeButton } from './SwitchThemeButton';
import { ThemeProvider } from '../../../contexts/ThemeProvider';

Object.defineProperty(window, 'matchMedia', {
  writable: true,
  value: jest.fn().mockImplementation((query) => ({
    matches: false,
    media: query,
    onchange: null,
    addEventListener: jest.fn(),
    removeEventListener: jest.fn(),
    dispatchEvent: jest.fn()
  }))
});

describe('SwitchThemeButton', () => {
  it('should switch the theme between light and dark', () => {
    const component = render(
      <ThemeProvider>
        <SwitchThemeButton />
      </ThemeProvider>
    );

    expect(screen.getByTestId('moon-icon')).toBeInTheDocument();

    expect(component).toMatchSnapshot();

    fireEvent.click(screen.getByRole('button'));

    expect(screen.getByTestId('sun-icon')).toBeInTheDocument();

    expect(component).toMatchSnapshot();
  });
});
