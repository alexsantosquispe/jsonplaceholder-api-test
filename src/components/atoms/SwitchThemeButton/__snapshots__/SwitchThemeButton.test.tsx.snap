// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SwitchThemeButton should switch the theme between light and dark 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <button
        aria-label="dark mode icon"
        class="transition-color cursor-pointer rounded-md p-4 text-black duration-150 ease-in-out hover:bg-gray-100 md:p-2 dark:text-white dark:hover:bg-white/10"
      >
        <svg
          class="size-5"
          data-testid="moon-icon"
          fill="none"
          stroke="currentColor"
          stroke-width="1.5"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M21.752 15.002A9.72 9.72 0 0 1 18 15.75c-5.385 0-9.75-4.365-9.75-9.75 0-1.33.266-2.597.748-3.752A9.753 9.753 0 0 0 3 11.25C3 16.635 7.365 21 12.75 21a9.753 9.753 0 0 0 9.002-5.998Z"
            stroke-linecap="round"
            stroke-linejoin="round"
          />
        </svg>
      </button>
    </div>
  </body>,
  "container": <div>
    <button
      aria-label="dark mode icon"
      class="transition-color cursor-pointer rounded-md p-4 text-black duration-150 ease-in-out hover:bg-gray-100 md:p-2 dark:text-white dark:hover:bg-white/10"
    >
      <svg
        class="size-5"
        data-testid="moon-icon"
        fill="none"
        stroke="currentColor"
        stroke-width="1.5"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M21.752 15.002A9.72 9.72 0 0 1 18 15.75c-5.385 0-9.75-4.365-9.75-9.75 0-1.33.266-2.597.748-3.752A9.753 9.753 0 0 0 3 11.25C3 16.635 7.365 21 12.75 21a9.753 9.753 0 0 0 9.002-5.998Z"
          stroke-linecap="round"
          stroke-linejoin="round"
        />
      </svg>
    </button>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`SwitchThemeButton should switch the theme between light and dark 2`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <button
        aria-label="light mode icon"
        class="transition-color cursor-pointer rounded-md p-4 text-black duration-150 ease-in-out hover:bg-gray-100 md:p-2 dark:text-white dark:hover:bg-white/10"
      >
        <svg
          class="size-5"
          data-testid="sun-icon"
          fill="none"
          stroke="currentColor"
          stroke-width="1.5"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M12 3v2.25m6.364.386-1.591 1.591M21 12h-2.25m-.386 6.364-1.591-1.591M12 18.75V21m-4.773-4.227-1.591 1.591M5.25 12H3m4.227-4.773L5.636 5.636M15.75 12a3.75 3.75 0 1 1-7.5 0 3.75 3.75 0 0 1 7.5 0Z"
            stroke-linecap="round"
            stroke-linejoin="round"
          />
        </svg>
      </button>
    </div>
  </body>,
  "container": <div>
    <button
      aria-label="light mode icon"
      class="transition-color cursor-pointer rounded-md p-4 text-black duration-150 ease-in-out hover:bg-gray-100 md:p-2 dark:text-white dark:hover:bg-white/10"
    >
      <svg
        class="size-5"
        data-testid="sun-icon"
        fill="none"
        stroke="currentColor"
        stroke-width="1.5"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M12 3v2.25m6.364.386-1.591 1.591M21 12h-2.25m-.386 6.364-1.591-1.591M12 18.75V21m-4.773-4.227-1.591 1.591M5.25 12H3m4.227-4.773L5.636 5.636M15.75 12a3.75 3.75 0 1 1-7.5 0 3.75 3.75 0 0 1 7.5 0Z"
          stroke-linecap="round"
          stroke-linejoin="round"
        />
      </svg>
    </button>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
